Security_Engine/Compute/CameraFieldOfView.cs

```diff
@@ -0,0 +1,351 @@
+        private static bool IsInsideObstacle(this Line line, List<Polyline> obstacles)
+        {
+            foreach (Polyline obstacle in obstacles)
+            {
+                foreach (Line obstLine in obstacle.SubParts())
+                {
+                    if (obstLine.IsOnCurve(line.Start) && obstLine.IsOnCurve(line.End))
+                        return false;
+                }
+                if (obstacle.IsContaining(new List<Point> { line.Centroid() }, true))
+                {
+                    return true;
+                }
+            }
+            return false;
+        }
+
+        /***************************************************/
+
+        [Description("Create visible line from rayline passing through obstacles.")]
+        [Input("obstacles", "Polyline objects that are needed to create visible line.")]
+        [Output("visibleLine", "Line that is visible for the camera view.")]
+        private static Line VisibleLine(this Line rayLine, List<Polyline> obstacles)
+        {
+            List<Line> splitLines = rayLine.SplitLinesInObstacles(obstacles);
+            List<Line> outsideLines = new List<Line>();
+            foreach (Line splitLine in splitLines)
+            {
+                if (!splitLine.IsInsideObstacle(obstacles))
+                    outsideLines.Add(splitLine);
+            }
+            //create visible line
+            Line visibleLine = new Line();
+            if (outsideLines.Count == 1)
+                visibleLine = outsideLines[0];
+            else
+            {
+                Point startPoint = outsideLines[0].Start;
+                Point endPoint = outsideLines[0].End;
+                for (int i = 1; i < outsideLines.Count; i++)
+                {
+                    if (endPoint.IsEqual(outsideLines[i].Start))
+                    {
+                        startPoint = outsideLines[0].End;
+                        endPoint = outsideLines[i].End;
+                    }
+                }
+                visibleLine = new Line { Start = startPoint, End = endPoint };
+            }
+
+            return visibleLine;
+        }
+
+        /***************************************************/
+
+        [Description("Create dictionary from line and obstacle.")]
+        [Input("line", "Line to create dictionary for.")]
+        [Input("obstacles", "Polyline objects that represents obstacles for lines.")]
+        [Output("dictionary", "Dictionary for lines and obstacles.")]
+        private static Dictionary<Line, Polyline> LineObstacleDictionary(this Line line, List<Polyline> obstacles)
+        {
+            Dictionary<Line, Polyline> lineObstDict = new Dictionary<Line, Polyline>();
+            if (obstacles.Count == 0)
+            {
+                lineObstDict.Add(line, null);
+            }
+            for (int i = 0; i < obstacles.Count; i++)
+            {
+                if (obstacles[i].IsContaining(new List<Point> { line.End }))
```

Comment:
As above, `IsContaining` is heavy, so worth trying to minimise the number of calls

Reply:
Moved to issue #2890.